atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/QosCassandraReadEteTest.java

```diff
@@ -0,0 +1,147 @@
+import com.palantir.atlasdb.todo.TodoSchema;
+import com.palantir.common.base.BatchingVisitable;
+
+public class QosCassandraReadEteTest extends QosCassandraEteTestSetup {
+    private static final int ONE_TODO_SIZE_IN_BYTES = 1050;
+
+    @BeforeClass
+    public static void createTransactionManagerAndWriteData() {
+        ensureTransactionManagerIsCreated();
+        writeNTodosOfSize(200, 1_000);
+        serializableTransactionManager.close();
+    }
+
+    @Test
+    public void shouldBeAbleToReadSmallAmountOfBytesIfDoesNotExceedLimit() {
+        assertThat(readOneBatchOfSize(1)).hasSize(1);
+    }
+
+    @Test
+    public void shouldBeAbleToReadSmallAmountOfBytesSeriallyIfDoesNotExceedLimit() {
+        IntStream.range(0, 50).forEach(i -> assertThat(readOneBatchOfSize(1)).hasSize(1));
+    }
+
+    @Test
+    public void shouldBeAbleToReadLargeAmountsExceedingTheLimitFirstTime() {
+        assertThat(readOneBatchOfSize(12)).hasSize(12);
+    }
+
+    @Test
+    public void shouldBeAbleToReadLargeAmountsExceedingTheLimitSecondTimeWithSoftLimiting() {
+        assertThat(readOneBatchOfSize(12)).hasSize(12);
+        // The second read might actually be faster as the transaction/metadata is cached
+        assertThat(readOneBatchOfSize(12)).hasSize(12);
+    }
+
+    @Test
+    public void shouldNotBeAbleToReadLargeAmountsIfSoftLimitSleepWillBeMoreThanConfiguredBackoffTime() {
+        assertThatThrownBy(() -> readOneBatchOfSize(200))
+                .isInstanceOf(RateLimitExceededException.class)
+                .hasMessage("Rate limited. Available capacity has been exhausted.");
+    }
+
+    @Test
+    public void readRateLimitShouldBeRespectedByConcurrentReadingThreads() throws InterruptedException {
+        int numThreads = 5;
+        int numReadsPerThread = 10;
+        ExecutorService executorService = Executors.newFixedThreadPool(numThreads);
+        List<Future<List<Todo>>> futures = new ArrayList<>(numThreads);
+
+        long start = System.nanoTime();
+        IntStream.range(0, numThreads).forEach(i ->
+                futures.add(executorService.submit(() -> {
+                    List<Todo> results = new ArrayList<>(numReadsPerThread);
+                    IntStream.range(0, numReadsPerThread)
+                            .forEach(j -> results.addAll(readOneBatchOfSize(1)));
+                    return results;
+                })));
+        executorService.shutdown();
+        Preconditions.checkState(executorService.awaitTermination(30L, TimeUnit.SECONDS),
+                "Read tasks did not finish in 30s");
+        long timeTakenToReadInSeconds = TimeUnit.NANOSECONDS.toSeconds(System.nanoTime() - start);
+
+        assertThatAllReadsWereSuccessful(futures, numReadsPerThread);
+        long actualReadRate = (numThreads * numReadsPerThread * ONE_TODO_SIZE_IN_BYTES) / timeTakenToReadInSeconds;
+        assertThat(actualReadRate).isLessThan(readBytesPerSecond + (readBytesPerSecond / 10 /* to allow burst time */));
```

Comment:
The bursty-ness allows us to save up to five extra seconds of permits, so I think we can bound this by `readBytesPerSecond * (time + 5) / time` rather than have the 1/10th fudge factor.

In general I'd prefer if we used doubles for the math above though, to avoid losses of precision with integer division

Reply:
Done